File: ios/Runner/Security/AppSealingThreatBridge.swift
import Flutter
import Foundation

final class AppSealingThreatBridge {
    static let shared = AppSealingThreatBridge()

    private let killDelay: TimeInterval = 3.0
    private var eventSink: FlutterEventSink?
    private var acknowledgedCodes = Set<Int>()

    private init() {}

    func setEventSink(_ sink: FlutterEventSink?) {
        eventSink = sink
    }

    func onThreatDetected(
        code: Int,
        message: String,
        category: String,
        willKillApp: Bool,
        extra: [String: Any] = [:]
    ) {
        emitThreatPayload(
            threatCode: extra["threat_code"] as? String,
            code: code,
            message: message,
            category: category,
            severity: extra["severity"] as? String,
            willKillApp: willKillApp,
            extra: extra
        )

        if willKillApp {
            DispatchQueue.main.asyncAfter(deadline: .now() + killDelay) { [weak self] in
                guard let self else { return }
                if !self.acknowledgedCodes.contains(code) {
                    exit(0)
                }
            }
        }
    }

    func onDoveRunnerThreatDiscovered(threatCode: String, timestampMillis: Int64) {
        let info = DoveRunnerThreatCatalog.resolve(threatCode)
        let numericCode = DoveRunnerThreatCatalog.parseNumericCode(info.threatCode)
        let detectedAtMillis = timestampMillis > 0 ? timestampMillis : Int64(Date().timeIntervalSince1970 * 1000)

        emitThreatPayload(
            threatCode: info.threatCode,
            code: numericCode,
            message: info.message,
            category: info.category,
            severity: info.severity,
            willKillApp: false,
            extra: [
                "source": "doverunner_callback",
                "raw_threat_code": threatCode,
                "timestamp_millis": detectedAtMillis,
            ],
            detectedAtMillis: detectedAtMillis
        )
    }

    private func emitThreatPayload(
        threatCode: String?,
        code: Int,
        message: String,
        category: String,
        severity: String?,
        willKillApp: Bool,
        extra: [String: Any],
        detectedAtMillis: Int64? = nil
    ) {
        let detectedAt: String
        if let millis = detectedAtMillis {
            detectedAt = ISO8601DateFormatter().string(from: Date(timeIntervalSince1970: TimeInterval(millis) / 1000))
        } else {
            detectedAt = ISO8601DateFormatter().string(from: Date())
        }

        var payload: [String: Any] = [
            "code": code,
            "message": message,
            "category": category,
            "platform": "ios",
            "will_kill_app": willKillApp,
            "detected_at": detectedAt,
            "extra": extra,
        ]
        if let threatCode {
            payload["threat_code"] = threatCode
        }
        if let severity {
            payload["severity"] = severity
        }

        DispatchQueue.main.async { [weak self] in
            self?.eventSink?(payload)
        }
    }

    func acknowledgeThreat(code: Int, reported: Bool) -> Bool {
        acknowledgedCodes.insert(code)
        return true
    }

    func isAppSealingActive() -> Bool {
        NSClassFromString("AppSealingCore") != nil
    }

    func handleMethodCall(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        switch call.method {
        case "acknowledgeThreat":
            guard let args = call.arguments as? [String: Any],
                  let code = args["code"] as? Int else {
                result(false)
                return
            }
            let reported = args["reported"] as? Bool ?? false
            result(acknowledgeThreat(code: code, reported: reported))

        case "isAppSealingActive":
            result(isAppSealingActive())

        default:
            result(FlutterMethodNotImplemented)
        }
    }
}

final class AppSealingThreatStreamHandler: NSObject, FlutterStreamHandler {
    func onListen(withArguments arguments: Any?, eventSink events: @escaping FlutterEventSink) -> FlutterError? {
        AppSealingThreatBridge.shared.setEventSink(events)
        return nil
    }

    func onCancel(withArguments arguments: Any?) -> FlutterError? {
        AppSealingThreatBridge.shared.setEventSink(nil)
        return nil
    }
}

File: android/app/src/main/kotlin/com/miftah/siap/security/AppSealingThreatBridge.kt
package com.miftah.siap.security

import android.os.Handler
import android.os.Looper
import io.flutter.plugin.common.EventChannel
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import java.time.Instant
import java.util.concurrent.ConcurrentHashMap

/**
 * Bridge DoveRunner threat broadcast → Flutter EventChannel → Dio → REST API.
 */
object AppSealingThreatBridge {
    private var eventSink: EventChannel.EventSink? = null
    private val mainHandler = Handler(Looper.getMainLooper())
    private val acknowledged = ConcurrentHashMap.newKeySet<Int>()
    private val recentBroadcastKeys = ConcurrentHashMap.newKeySet<String>()

    fun setEventSink(sink: EventChannel.EventSink?) {
        eventSink = sink
    }

    /**
     * Entry point dari [ThreatReceiver] sesuai DoveRunner Monitoring Mode.
     * Broadcast hanya mengirim threat_code + timestamp; app tidak di-kill.
     */
    fun onDoveRunnerThreatDiscovered(threatCode: String, timestampMillis: Long) {
        val info = DoveRunnerThreatCatalog.resolve(threatCode)
        val numericCode = DoveRunnerThreatCatalog.parseNumericCode(info.threatCode)
        val detectedAtMillis = if (timestampMillis > 0L) timestampMillis else System.currentTimeMillis()

        // DoveRunner mendeteksi berkala (~5 detik). Dedup per bucket 5 detik.
        val dedupeKey = "${info.threatCode}-${detectedAtMillis / 5_000L}"
        if (!recentBroadcastKeys.add(dedupeKey)) {
            return
        }

        val payload = linkedMapOf<String, Any?>(
            "threat_code" to info.threatCode,
            "code" to numericCode,
            "message" to info.message,
            "category" to info.category,
            "severity" to info.severity,
            "platform" to "android",
            "will_kill_app" to false,
            "detected_at" to Instant.ofEpochMilli(detectedAtMillis).toString(),
            "extra" to mapOf(
                "source" to "doverunner_broadcast",
                "raw_threat_code" to threatCode,
                "timestamp_millis" to detectedAtMillis,
            ),
        )

        mainHandler.post {
            eventSink?.success(payload)
        }
    }

    fun acknowledgeThreat(code: Int, reported: Boolean): Boolean {
        acknowledged.add(code)
        return true
    }

    fun isAppSealingActive(): Boolean {
        return isDoveRunnerSealed() || isLegacyAppSealingPresent()
    }

    private fun isDoveRunnerSealed(): Boolean {
        return try {
            Class.forName("com.doverunner.aos.security.AppSealingAlertDialog")
            true
        } catch (_: ClassNotFoundException) {
            false
        }
    }

    private fun isLegacyAppSealingPresent(): Boolean {
        return try {
            Class.forName("com.inka.appsealing.AppSealingAlertDialog")
            true
        } catch (_: ClassNotFoundException) {
            false
        }
    }

    fun handleMethodCall(call: MethodCall, result: MethodChannel.Result) {
        when (call.method) {
            "acknowledgeThreat" -> {
                val code = call.argument<Int>("code") ?: 0
                val reported = call.argument<Boolean>("reported") ?: false
                result.success(acknowledgeThreat(code, reported))
            }
            "isAppSealingActive" -> result.success(isAppSealingActive())
            else -> result.notImplemented()
        }
    }
}
